druid/src/widget/images.rs

```diff
@@ -0,0 +1,259 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An Image widget.
+//! Please consider using Image and the Image wideget as it scales much better.
+
+use std::error::Error;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use crate::{
+    Affine, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx,
+    PaintCtx, Point, Rect, RenderContext, Size, UpdateCtx, Widget,
+};
+
+use crate::piet::{ImageFormat, InterpolationMode};
+
+use image;
+
+#[derive(PartialEq)]
+pub enum FillStrat {
+    Contain,
+    Cover,
+    Fill,
+    FitHeight,
+    FitWidth,
+    None,
+    ScaleDown,
+}
+
+pub enum FloatVert {
+    Top,
+    Center,
+    Bottom,
+}
+
+pub enum FloatHos {
+    Left,
+    Center,
+    Right,
+}
+
+fn get_scale_offset(
+    pairent: Size,
+    fit_box: Size,
+    fit_type: &FillStrat,
+    _fhos: &FloatHos,
+    _fvert: &FloatVert,
+) -> (Point, Point) {
+    let scalex = pairent.width / fit_box.width;
+    let scaley = pairent.height / fit_box.height;
+
+    let scale: Point = match fit_type {
+        FillStrat::Contain => {
+            let scale = scalex.min(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Cover => {
+            let scale = scalex.max(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Fill => Point {
+            x: scalex,
+            y: scaley,
+        },
+        FillStrat::FitHeight => Point {
+            x: scaley,
+            y: scaley,
+        },
+        FillStrat::FitWidth => Point {
+            x: scalex,
+            y: scalex,
+        },
+        FillStrat::ScaleDown => {
+            let scale = scalex.min(scaley).min(1.0);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::None => Point { x: 1.0, y: 1.0 },
+    };
+
+    let origin_x = (pairent.width - (fit_box.width * scale.x)) / 2.0;
+    let origin_y = (pairent.height - (fit_box.height * scale.y)) / 2.0;
+    let origin = Point::new(origin_x, origin_y);
+
+    (scale, origin)
+}
+
+/// A widget that renders a Images
+pub struct Image<T> {
+    image_data: ImageData,
+    phantom: PhantomData<T>,
+
+    vert: FloatVert,
+    hoz: FloatHos,
+    fill: FillStrat,
+}
+
+impl<T: Data> Image<T> {
+    /// Create an Image-drawing widget from ImageData.
+    ///
+    /// The Image will scale to fit its box constraints.
+    pub fn new(image_data: ImageData, fillstr: FillStrat) -> impl Widget<T> {
```

Comment:
is there a reasonable default `FIllStrat`? If there is I might consider having `new` only take `ImageData`, and then adding a builder method to set the `FillStrat` for people that need it.

Reply:
As set_fill takes a `mut` of self i have had to add a `mut` to the let of the image widget.

dose this seem fair or should i tweak the `set_fill`?

Reply:
What I would do is have a separate function that is used at build time, like

```rust
...
pub fn fill_mode(mut self, mode: FillStrat) -> Self {
    self.fill = fillstr;
    self
}
```

whatever name is fine, but not `set_`; I would reserve that for functions that modify the value after construction.